homeassistant/components/mqtt/__init__.py

```diff
@@ -370,18 +389,30 @@ def async_publish_service(call):
         msg_topic = call.data[ATTR_TOPIC]
         payload = call.data.get(ATTR_PAYLOAD)
         payload_template = call.data.get(ATTR_PAYLOAD_TEMPLATE)
+        payload_file_path = call.data.get(ATTR_PAYLOAD_FILE_PATH)
         qos = call.data[ATTR_QOS]
         retain = call.data[ATTR_RETAIN]
-        try:
-            if payload_template is not None:
+        if payload_template is not None:
+            try:
                 payload = \
                     template.Template(payload_template, hass).async_render()
-        except template.jinja2.TemplateError as exc:
-            _LOGGER.error(
-                "Unable to publish to '%s': rendering payload template of "
-                "'%s' failed because %s",
-                msg_topic, payload_template, exc)
-            return
+            except template.jinja2.TemplateError as exc:
+                _LOGGER.error(
+                    "Unable to publish to '%s': rendering payload template of "
+                    "'%s' failed because %s",
+                    msg_topic, payload_template, exc)
+                return
+        elif payload_file_path is not None:
```

Comment:
Paulus, as I explained before, I was building on top of an existing service that doesn't allow binary payloads.
I tried to add a feature that is provided by the MQTT protocol but not by the implementation of this service. Admittedly, my first attempt was flawed, I hadn't thought about the security considerations. But after? How sending a file that is already available to the frontend (being in the `www` dir) could be possibly considered a security flaw? I even put in a path traversal safeguard...
And the argument that this is not something to be put in a base component, well, that is valid for the whole service, isn't it? I say this because there's no class "A" payload (strings) and class "B" payloads (the rest). They are all the same, actually they are all binary as you know,  and so they should be treated.

So the problem is not passing a file, but not being able to send binary data to the `publish` service and have them published _as is_

I reckon you disagree, well, no problem: I'll write an mqtt notifier component ;)